### ISSUE TITLE
After the update to Lima 0.14 all containers, images and volumes are gone

### ISSUE THREAD

**Issue by rpkoller** [abiosoft/colima#506]
### Description

Few days ago Lima 0.14 was released. since no extra steps were required in the changelog (except the support of virtiofs) i was adventurous and update lima to 0.14. i am using colima with the docker runtime to run DDEV in. I am using homebrew and therefore did a `brew upgrade lima`.  colima started without a complaint after that. the odd part when starting a project in ddev the first time i got. 

```
$> ddev start
Network ddev_default created 
Starting projectbrowser... 
v1.21.4: Pulling from drud/ddev-webserver
1ef0eb6dda52: Pull complete 
Digest: sha256:6a2e66d09a166e36179311f92a8ae0e8baa4fbed43fadf31be3c62995ead98cb
Status: Downloaded newer image for drud/ddev-webserver:v1.21.4
docker.io/drud/ddev-webserver:v1.21.4
stable: Pulling from library/busybox
b2bcf7a3443b: Pull complete 
Digest: sha256:3b3128d9df6bbbcc92e2358e596c9fbd722a437a62bafbc51607970e9e3b8869
Status: Downloaded newer image for busybox:stable
docker.io/library/busybox:stable
v1.21.4: Pulling from drud/ddev-ssh-agent
6064e7e5b6af: Pull complete 
bf273825688f: Pull complete 
7c56dd07c5c1: Pull complete 
4f4fb700ef54: Pull complete 
0e79da486cab: Pull complete 
b9227daed8ed: Pull complete 
```
all images got downloaded again.  and that part in the startup output struck my eye: 

```
Your docker install has only 2136836 available disk space, less than 5000000 warning level (48% used). Please increase disk image size. 
Volume "ddev-ssh-agent_dot_ssh"  Created 
Volume "ddev-ssh-agent_socket_dir"  Created 
Container ddev-ssh-agent  Started 
```
i've checked if there is a new docker context created maybe but no: 

```
$> docker context ls
NAME       DESCRIPTION                               DOCKER ENDPOINT                                     KUBERNETES ENDPOINT   ORCHESTRATOR
colima *   colima                                    unix:///Users/rkoller/.colima/default/docker.sock                         
default    Current DOCKER_HOST based configuration   unix:///var/run/docker.sock                                               swarm
```
so all the containers, images and volumes are gone. but the odd detail is the following. according to:  
```
$> docker system df   
TYPE            TOTAL     ACTIVE    SIZE      RECLAIMABLE
Images          10        5         2.433GB   1.996GB (82%)
Containers      5         5         54.38MB   0B (0%)
Local Volumes   7         5         420.2MB   0B (0%)
Build Cache     52        0         2.116kB   2.116kB
```
all the images, containers, volumes and caches have approximately 5gb all together.  powered off ddev and restarted a project and the available disk space warning still exists: 

```
$> ddev start
Network ddev_default created 
Starting projectbrowser... 
Your docker install has only 544552 available disk space, less than 5000000 warning level (87% used). Please increase disk image size. 
```
so not sure if the old containers/images and so on are still in place but unavailable or if something else went wrong as well. 

### Version

Colima Version:
```
$> colima version
colima version 0.4.6
git commit: 10377f3a20c2b0f7196ad5944264b69f048a3d40

runtime: docker
arch: aarch64
client: v20.10.21
server: v20.10.11
```
Lima Version:
```
$> limactl --version
limactl version 0.14.0
```
Qemu Version:
$> qemu-img --version
qemu-img version 7.1.0
Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers

I've installed colima via homebrew on a mbp (m1pro)

### Operating System

- [ ] macOS Intel
- [X] macOS M1
- [ ] Linux

### Reproduction Steps

1. upgrade lima to 0.14

p.s. @rfay tested upgrading to lima 0.14 with colima HEAD and he confirms, same outcome.

**Comment by rfay** [comment]
I found that upgrading with colima HEAD also blew away all images/containers/volumes.

**Comment by abiosoft** [comment]
Thanks for reporting this, it was somehow not properly considered. I admit that there needs to be more documentation in this regard to prepare users for the transition.

Yeah, it is due to the change in the underlying virtual machine tool from QEMU to Virtualization.Framework. To transition from one to the other would behave like a reset i.e. loss of existing data in the VMs.

I think Colima should simply re-use QEMU for existing VMs to prevent loss of data, and Virtualization.Framework for new instances. That should provide a better experience.

**Comment by rfay** [comment]
A related note: Even though all images were gone, the disk still was all used up. I had to `colima delete` and then recreate to get enough disk to work.

**Comment by rpkoller** [comment]
i agree staying with the virtual machine tool the VM is created in is definitely the desirable thing and use virtualization.framework as the default sounds as a reasonable choice as well. 

the only thing i am not sure is about your statement that changing from qemu to virtualization framework would behave like a "reset". my VM has 100gb. after the update the assets had ~5gb but i get the available disk space warning that over 50gb are already used. so it seems it isnt completely reset? or something else happens.

**Comment by abiosoft** [comment]
@rfay @rpkoller what happened is that the way the disk partitioning works for either platforms is different. The existing disk is not deleted, but rather the free space is being utilised and re-partitioned.

Technically, the data is still there but it is no longer accessible.

**Comment by balajiv113** [comment]
@abiosoft 
Does it makes sense to provide a support for migration of disks between vmTypes ??
vz - requires the diffdisk to be of raw format
qemu - accepts qcow2 as diffdisk format

Within lima, qemu-img is used to convert diffdisk from qcow2 to raw format. Doing the reverse should also be possible

**Comment by abiosoft** [comment]
@balajiv113 considering that it may provide the ability to transition between vmTypes without losing user data, it sounds like a welcoming idea.

But at the same time, I am considering the practical use-case and I am leaning towards it being rarely utilised.

**Comment by abiosoft** [comment]
> To confirm, this issue only happens to colima@HEAD ([40bf7bf](https://github.com/abiosoft/colima/commit/40bf7bf79d4b1cdad801af37ca4b71e325d32c29)) users?
> 
> IIUC, colima v0.4.6 (and older) users, and vanilla Lima users are not affected, as it does not contain the code to change the default `vmType` to `vz`?

Exactly 💯

**Comment by AkihiroSuda** [comment]
> Migration

Migration can be dangerous, so I think the default behavior should be to just refuse booting when qcow2 is passed to the vz driver.

**Comment by balajiv113** [comment]
> But at the same time, I am considering the practical use-case and I am leaning towards it being rarely utilised.

True 👍  For this default transition alone it will be helpful.

This also won't work if people started a emulated vm (running amd64 in arm64 ) as Virtualization.Framework doesn't support that.

**Comment by rfay** [comment]
> To confirm, this issue only happens to colima@HEAD (https://github.com/abiosoft/colima/commit/40bf7bf79d4b1cdad801af37ca4b71e325d32c29) users?

@AkihiroSuda this is *not* correct. This affects current stable release of Colima, because colima depends on Lima via homebrew, so the homebrew update changes the back-end and breaks users of colima without using HEAD.

**Comment by abiosoft** [comment]
> @AkihiroSuda this is *not* correct. This affects current stable release of Colima, because colima depends on Lima via homebrew, so the homebrew update changes the back-end and breaks users of colima without using HEAD.

Really, is this the case for you? From my understanding and observation so far, it will only affect those following (and testing) the virtiofs development and building Colima from source. Not the average user.

**Comment by rpkoller** [comment]
I've reported the issue using colima 0.4.6 updating lima to 0.14 and i ran into the issue as described in the issue summary

**Comment by AkihiroSuda** [comment]
Sorry, I confirmed that the issue is actually reproducible with colima v0.4.6 (tested by running `docker pull hello-world`) 🙇 
So, this is irrelevant to `vz`.

But this doesn't seem reproducible with vanilla Lima. Tested the `default` template (`nerdctl pull hello-world`) and `alpine` template (`sudo touch /var/lib/foo`, as the vanilla `alpine` template does not contain container runtime).

**Comment by AkihiroSuda** [comment]
BTW apparently the data are not removed, they are just somehow unmounted in Lima v0.14.

The data can be restored by rolling back Lima to v0.13.
```shell
brew uninstall lima
git clone https://github.com/lima-vm/lima.git
git checkout v0.13.0
make
make install
```

**Comment by AkihiroSuda** [comment]
Fix is in progress
- https://github.com/lima-vm/lima/pull/1243

**Comment by rfay** [comment]
Thanks!

**Comment by AkihiroSuda** [comment]
This was fixed in Lima v0.14.1.

Already available for `brew install --HEAD lima`.
Will be available for `brew install lima` soon:
- https://github.com/Homebrew/homebrew-core/pull/118132

**Comment by rfay** [comment]
You're awesome!
